### PR TITLE
pyarrow: Support date32[day] and date64[ms] dtypes in pandas objects

### PR DESCRIPTION
**Pull request by weiji14**
**Description of proposed changes**

Handle date columns in pandas.DataFrame with pyarrow dtypes like `date32[day][pyarrow]` or `date64[ms][pyarrow]`.

This is implemented by modifying the [`vectors_to_arrays`](https://github.com/GenericMappingTools/pygmt/blob/v0.10.0/pygmt/clib/conversion.py#L127) conversion function to convert date32/date64 dtypes to numpy.datetime64.

An altenative implementation would be to handle pyarrow date32/date64 dtypes natively without converting, but this might require more work and discussion at #2800.

Note:
- Unlike numpy which only has datetime64, PyArrow has separate [Date types (date64/date32)](https://arrow.apache.org/docs/python/pandas.html#date-types) and [Time types (time64)](https://arrow.apache.org/docs/python/pandas.html#time-types).

References:
- https://arrow.apache.org/docs/python/pandas.html#date-types
- https://arrow.apache.org/docs/13.0/python/generated/pyarrow.date32.html
- https://arrow.apache.org/docs/13.0/python/generated/pyarrow.date64.html

Part of #2800, related to #242, extends #464

**Reminders**

- [x] Run `make format` and `make check` to make sure the code follows the style guide.
- [x] Add tests for new features or tests that would have caught the bug that you're fixing.
- [ ] Add new public functions/methods/classes to `doc/api/index.rst`.
- [x] Write detailed docstrings for all functions/methods.
- [ ] If wrapping a new module, open a 'Wrap new GMT module' issue and submit reasonably-sized PRs.
- [ ] If adding new functionality, add an example to docstrings or tutorials.
- [ ] Use underscores (not hyphens) in names of Python files and directories.

**Slash Commands**

You can write slash commands (`/command`) in the first line of a comment to perform
specific operations. Supported slash commands are:

- `/format`: automatically format and lint the code
- `/test-gmt-dev`: run full tests on the latest GMT development version

### CODE REVIEW

File: pygmt/clib/conversion.py [hunk 1]
```diff
@@ -162,11 +162,43 @@ def vectors_to_arrays(vectors):
...
+    >>> arrays = vectors_to_arrays(vectors)
+    >>> all(a.flags.c_contiguous for a in arrays)
+    True
+    >>> all(isinstance(a, np.ndarray) for a in arrays)
+    True
+    >>> all(isinstance(a.dtype, np.dtypes.DateTime64DType) for a in arrays)
+    True
     """
-    arrays = [as_c_contiguous(np.asarray(i)) for i in vectors]
+    arrays = []
```
Comment (seisman) [inline pygmt/clib/conversion.py:191]: I haven't tried it, but maybe the following code is simpler:
```
dtypes = {
    "date32[day][pyarrow]": np.datetime64,
    "date64[ms][pyarrow]": np.datetime64,
}
arrays = []
for vector in vectors:
    vec_dtype = str(getattr(vector, "dtype", ""))
    array = np.asarray(dtypes.get(vec_dtype, None))
    arrays.append(as_c_contiguous(array))
```

Comment (weiji14) [reply pygmt/clib/conversion.py:195]: Oh nice, that's much simpler and makes it easier to extend to other dtypes in the future! Done at commit de08508d82fe2e5e79255542ca65193010320699.

**Review by seisman — APPROVED** [review]
Looks good to me, but need to mark the PR as ready for review and merge the main branch to make sure all tests pass.

**Comment by weiji14** [comment]
I'm still debating on whether to add pyarrow into `test_plot_datetime` that was added in #464: https://github.com/GenericMappingTools/pygmt/blob/20054a13627729412a14c9562595237d384d143a/pygmt/tests/test_plot.py#L409-L452

**Comment by weiji14** [comment]
> I'm still debating on whether to add pyarrow into `test_plot_datetime` that was added in #464:

On second thought, I might do this in a follow-up PR. PyArrow also has time32/time64 types, so I might add a `test_plot_datetime_pyarrow` unit test that reuses `test_plot_datetime`'s baseline image, but using pyarrow inputs instead.

**Pull request merged**